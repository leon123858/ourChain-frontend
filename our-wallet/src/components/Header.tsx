import {FaBeer} from "react-icons/fa";
import {AiOutlineBars} from "react-icons/ai";
import {Drawer, Flex, Menu} from 'antd';
import {Context, JSX, useContext, useState} from "react";
import {UserContext} from "../utils/context/user.tsx";
import {signIn} from "../utils/userApiWrapper/firebaseAuth.ts";


const AppHeader = ({handleSignIn = signIn, context = UserContext, menu = <Menu/>}: {
    handleSignIn?: () => Promise<void>,
    context?: Context<{ isLogin: boolean, email: string, handleLogout: () => Promise<void> }>,
    menu?: JSX.Element
}) => {
    const {isLogin, email} = useContext(context);
    const [visible, setVisible] = useState(false);

    return (
        <div style={{width: "100vw", backgroundColor: "black", height: "8vh"}}>
            <Flex justify={"space-between"} style={{
                width: "100%",
                height: "100%",
                alignItems: "center",
                padding: "0 10px",
                fontSize: "25px",
            }}>
                <FaBeer/>
                {isLogin ?
                    <span className={"logoutBtn"}
                          style={{color: "whitesmoke", maxWidth: "30vw", overflow: "hidden", whiteSpace: "nowrap"}}
                    >
                        {email}
                    </span> : <button className={"loginBtn"} style={{color: "white"}} onClick={async () => {
                        await handleSignIn();
                    }}>
                        Login
                    </button>}
                <button onClick={() => {
                    setVisible(true)
                }}>
                    <AiOutlineBars/>
                </button>
            </Flex>
            <Drawer
                title="Menu"
                placement="right"
                open={visible}
                onClose={() => {
                    setVisible(false)
                }}
            >
                {menu}
            </Drawer>
        </div>
    );
};

export default AppHeader;