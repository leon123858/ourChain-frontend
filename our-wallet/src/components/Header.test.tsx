/**
 * @vitest-environment jsdom
 */

import renderer from 'react-test-renderer'
import {describe, expect, test} from "vitest";
import Header from "./Header.tsx";
import {createContext} from "react";

describe('Header', () => {
    const MockUserContext = createContext({
        isLogin: false,
        email: "",
        handleLogout: async () => {
        }
    });

    test('try login', () => {
        let spyTriggerTime = 0;
        const spyFunc = () => {
            spyTriggerTime += 1;
        }
        const contextValue = {
            isLogin: false,
            handleLogout: async () => {
            },
            email: "",
        }
        const component = renderer.create(
            <MockUserContext.Provider value={contextValue}>
                <Header handleSignIn={async () => {
                    spyFunc()
                }} context={MockUserContext}></Header>
            </MockUserContext.Provider>,
        )
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
        expect(tree).toBeDefined()

        const instance = component.root
        expect(instance).toBeDefined()

        const btn = instance.findByProps({className: 'loginBtn'})
        expect(btn?.children[0]).toBe('Login')
        for (let i = 0; i < 10; i++) {
            btn.props.onClick()
            expect(spyTriggerTime).toBe(i + 1)
        }
    })

    test('try logout', () => {
        let spyTriggerTime = 0;
        const spyFunc = () => {
            spyTriggerTime += 1;
        }
        const component = renderer.create(
            <MockUserContext.Provider value={{
                isLogin: true,
                handleLogout: async () => {
                    spyFunc()
                },
                email: "test@sample.com"
            }}>
                <Header handleSignIn={async () => {
                }} context={MockUserContext}></Header>
            </MockUserContext.Provider>
        )
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
        expect(tree).toBeDefined()

        const instance = component.root
        expect(instance).toBeDefined()

        const btn = instance.findByProps({className: 'logoutBtn'})
        expect(btn?.children[0]).toBe('test@sample.com')
    })
})
