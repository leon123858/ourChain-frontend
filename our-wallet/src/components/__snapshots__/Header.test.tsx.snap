// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > try login 1`] = `
<div
  style={
    {
      "backgroundColor": "black",
      "height": "8vh",
      "width": "100vw",
    }
  }
>
  <div
    className="ant-flex css-dev-only-do-not-override-2q8sxy ant-flex-justify-space-between"
    style={
      {
        "alignItems": "center",
        "fontSize": "25px",
        "height": "100%",
        "padding": "0 10px",
        "width": "100%",
      }
    }
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M368 96h-48V56c0-13.255-10.745-24-24-24H24C10.745 32 0 42.745 0 56v400c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24v-42.11l80.606-35.977C429.396 365.063 448 336.388 448 304.86V176c0-44.112-35.888-80-80-80zm16 208.86a16.018 16.018 0 0 1-9.479 14.611L320 343.805V160h48c8.822 0 16 7.178 16 16v128.86zM208 384c-8.836 0-16-7.164-16-16V144c0-8.836 7.164-16 16-16s16 7.164 16 16v224c0 8.836-7.164 16-16 16zm-96 0c-8.836 0-16-7.164-16-16V144c0-8.836 7.164-16 16-16s16 7.164 16 16v224c0 8.836-7.164 16-16 16z"
      />
    </svg>
    <button
      className="loginBtn"
      onClick={[Function]}
      style={
        {
          "color": "white",
        }
      }
    >
      Login
    </button>
    <button>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M912 192H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM104 228a56 56 0 1 0 112 0 56 56 0 1 0-112 0zm0 284a56 56 0 1 0 112 0 56 56 0 1 0-112 0zm0 284a56 56 0 1 0 112 0 56 56 0 1 0-112 0z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Header > try logout 1`] = `
<div
  style={
    {
      "backgroundColor": "black",
      "height": "8vh",
      "width": "100vw",
    }
  }
>
  <div
    className="ant-flex css-dev-only-do-not-override-2q8sxy ant-flex-justify-space-between"
    style={
      {
        "alignItems": "center",
        "fontSize": "25px",
        "height": "100%",
        "padding": "0 10px",
        "width": "100%",
      }
    }
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M368 96h-48V56c0-13.255-10.745-24-24-24H24C10.745 32 0 42.745 0 56v400c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24v-42.11l80.606-35.977C429.396 365.063 448 336.388 448 304.86V176c0-44.112-35.888-80-80-80zm16 208.86a16.018 16.018 0 0 1-9.479 14.611L320 343.805V160h48c8.822 0 16 7.178 16 16v128.86zM208 384c-8.836 0-16-7.164-16-16V144c0-8.836 7.164-16 16-16s16 7.164 16 16v224c0 8.836-7.164 16-16 16zm-96 0c-8.836 0-16-7.164-16-16V144c0-8.836 7.164-16 16-16s16 7.164 16 16v224c0 8.836-7.164 16-16 16z"
      />
    </svg>
    <button
      className="logoutBtn"
      onClick={[Function]}
      style={
        {
          "color": "whitesmoke",
          "maxWidth": "30vw",
          "overflow": "hidden",
          "whiteSpace": "nowrap",
        }
      }
    >
      Logout 
      test@sample.com
    </button>
    <button>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M912 192H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM104 228a56 56 0 1 0 112 0 56 56 0 1 0-112 0zm0 284a56 56 0 1 0 112 0 56 56 0 1 0-112 0zm0 284a56 56 0 1 0 112 0 56 56 0 1 0-112 0z"
        />
      </svg>
    </button>
  </div>
</div>
`;
